### DISCUSSION TITLE
How to train multiple LoRA adapters on the same base model concurrently.

### DISCUSSION THREAD

**Discussion by rajat95** [huggingface/peft#1315] [General]
I want to train 2 LoRA models in conjunction on my dataset. I don't want gradients from one model impact the other. However,  since the base model is the same I am confused if just setting adapter_name like below would achieve this purpose? 

```
        modelA = get_peft_model(model, lora_config, adapter_name="A")
        modelB = get_peft_model(model, profile_lora_config, adapter_name="B")
```

And then in the model forward
```
        outputA = modelA.forward(inputsA)
        outputB = modelB.forward(inputsB)
```

**Comment by BenjaminBossan** [comment]
Hi, this is not the way. Could you try something like this instead?

```python
model = get_peft_model(model, lora_config, adapter_name="A")
# add 2nd adapter, "A" is still the active adapter
model.add_adapter("B", profile_lora_config)
# now train adapter "A"
...
# switch to adapter "B" and train it
model.set_adapter("B")
...
```

**Comment by rajat95** [comment]
@BenjaminBossan Thanks for the reply. However adapter A and B are trained simultaneously since loss depends on both outputA and outputB. Is there a way to achieve that?

**Reply by BenjaminBossan** [reply]
You can activate multiple adapters at once:

```python
model.base_model.set_adapter(["A", "B"])
```

This way, you can create outputs with the model that only include adapter A, B, or both. Note that this won't work if you use `modules_to_save`.

**Comment by Weibin44** [comment]
What's the real function of `set_adapter`? Once we use `model.set_adapter("B")`, does it mean there's no influence of `adapter A` during the forward and backward? or just one of them? Thanks for your help in adavance!

**Reply by BenjaminBossan** [reply]
Yes, after setting `model.set_adapter("B")`, only the adapter called `"B"` will participate in the `forward` call of the model. As such, there should be no gradients for adapter `"A"` and thus its parameters are not updated.
